Add initial support for SPE brstack format

kaadam opened the pull request:
Perf will be able to report SPE branch events as similar as it does with LBR brstack.
Therefore we can utilize the existing LBR parsing process for SPE as well.

Example of the SPE brstack input format:
```bash
perf script -i perf.data -F pid,brstack --itrace=bl
```
```
  PID       FROM / TO / PREDICTED

16984  0x72e342e5f4/0x72e36192d0/M/-/-/11/RET/-
16984  0x72e7b8b3b4/0x72e7b8b3b8/PN/-/-/11/COND/-
16984  0x72e7b92b48/0x72e7b92b4c/PN/-/-/8/COND/-
16984  0x72eacc6b7c/0x760cc94b00/P/-/-/9/RET/-
16984  0x72e3f210fc/0x72e3f21068/P/-/-/4//-
16984  0x72e39b8c5c/0x72e3627b24/P/-/-/4//-
16984  0x72e7b89d20/0x72e7b92bbc/P/-/-/4/RET/-
```
SPE brstack mispredicted flag might be two characters long: `PN` or `MN`. Where `N` means the branch was marked as NOT-TAKEN. This event is only related to conditional instruction (conditional branch or compare-and-branch), it tells that failed its condition code check.

Perf with 'brstack' support for SPE is available here:
```
https://github.com/Leo-Yan/linux/tree/perf_arm_spe_branch_flags_v2
```

Example of useage with SPE perf data:
```bash
perf2bolt -p perf.data -o perf.fdata --spe BINARY
```

Capture standard SPE branch events with perf:
```bash
perf record -e 'arm_spe_0/branch_filter=1/u' -- BINARY
```

An unittest is also added to check parsing process of 'SPE brstack format'.

paschalis-mpeis committed:
[BOLT][AArch64] Introduce SPE mode in BasicAggregation

BOLT gains the ability to process branch target information generated by
Arm SPE data, using the `BasicAggregation` format.

Example usage is:
```bash
perf2bolt -p perf.data -o perf.boltdata --nl --spe BINARY
```

New branch data and compatibility:
---
SPE branch entries in perf data contain a branch pair (`IP` -> `ADDR`)
for the source and destination branches. DataAggregator processes those
by creating two basic samples. Any other event types will have `ADDR`
field set to `0x0`. For those a single sample will be created. Such
events can be either SPE or non-SPE, like `l1d-access` and `cycles`
respectively.

The format of the input perf entries is:
```
PID   EVENT-TYPE   ADDR   IP
```

When on SPE mode and:
- host is not `AArch64`, BOLT will exit with a relevant message
- `ADDR` field is unavailable, BOLT will exit with a relevant message
- no branch pairs were recorded, BOLT will present a warning

Examples of generating profiling data for the SPE mode:
---
Profiles can be captured with perf on AArch64 machines with SPE enabled.
They can be combined with other events, SPE or not.

Capture only SPE branch data events:
```bash
perf record -e 'arm_spe_0/branch_filter=1/u' -- BINARY
```

Capture any SPE events:
```bash
perf record -e 'arm_spe_0//u' -- BINARY
```

Capture any SPE events and cycles
```bash
perf record -e 'arm_spe_0//u' -e cycles:u -- BINARY
```

More filters, jitter, and specify count to control overheads/quality.
```bash
perf record -e 'arm_spe_0/branch_filter=1,load_filter=0,store_filter=0,jitter=1/u' -c 10007 -- BINARY
```

paschalis-mpeis committed:
clang-format fix

paschalis-mpeis committed:
Addressing reviewers (1)

paschalis-mpeis committed:
Addressing reviewers (2)

kaadam committed:
Add initial support for SPE brstack

Perf will be able to report SPE branch events as similar as it does
with LBR brstack.
Therefore we can utilize the existing LBR parsing process for SPE as well.

Example of the SPE brstack input format:
```bash
perf script -i perf.data -F pid,brstack --itrace=bl
```
```
---
PID    FROM         TO           PREDICTED
---
16984  0x72e342e5f4/0x72e36192d0/M/-/-/11/RET/-
16984  0x72e7b8b3b4/0x72e7b8b3b8/PN/-/-/11/COND/-
16984  0x72e7b92b48/0x72e7b92b4c/PN/-/-/8/COND/-
16984  0x72eacc6b7c/0x760cc94b00/P/-/-/9/RET/-
16984  0x72e3f210fc/0x72e3f21068/P/-/-/4//-
16984  0x72e39b8c5c/0x72e3627b24/P/-/-/4//-
16984  0x72e7b89d20/0x72e7b92bbc/P/-/-/4/RET/-
```
SPE brstack mispredicted flag might be two characters long: 'PN' or 'MN'.
Where 'N' means the branch was marked as NOT-TAKEN. This event is only related to
conditional instruction (conditional branch or compare-and-branch),
it tells that failed its condition code check.

Perf with 'brstack' support for SPE is available here:
```
https://github.com/Leo-Yan/linux/tree/perf_arm_spe_branch_flags_v2
```

Example of useage with SPE perf data:
```bash
perf2bolt -p perf.data -o perf.fdata --spe BINARY
```

Capture standard SPE branch events with perf:
```bash
perf record -e 'arm_spe_0/branch_filter=1/u' -- BINARY
```

An unittest is also added to check parsing process of 'SPE brstack format'.

paschalis-mpeis commented on bolt/test/perf2bolt/AArch64/perf2bolt-spe.test:
```diff
@@ -11,4 +11,4 @@ CHECK-SPE-NO-LBR: PERF2BOLT: Starting data aggregation job
 RUN: not perf2bolt -p %t.perf.data -o %t.perf.boltdata --spe %t.exe 2>&1 | FileCheck %s --check-prefix=CHECK-SPE-LBR
 
-CHECK-SPE-LBR: PERF2BOLT-ERROR: Arm SPE mode is combined only with BasicAggregation.
+CHECK-SPE-LBR: PERF2BOLT: spawning perf job to read SPE branch events
```
This test was not failing on my machine (ie not needing `not perf2bolt`), which made me realize that these tests might be fragile on buildbots and local environments. So I suggest using `perf record` for both non-lbr/lbr testing, and overriding the exit value with a sub shell like so:

```diff
-## Check that Arm SPE mode is available on AArch64 with BasicAggregation.
+## Check that Arm SPE mode is available on AArch64.
 
 REQUIRES: system-linux,perf,target=aarch64{{.*}}
 
 RUN: %clang %cflags %p/../../Inputs/asm_foo.s %p/../../Inputs/asm_main.c -o %t.exe
-RUN: touch %t.empty.perf.data
-RUN: perf2bolt -p %t.empty.perf.data -o %t.perf.boltdata --nl --spe --pa %t.exe 2>&1 | FileCheck %s --check-prefix=CHECK-SPE-NO-LBR
+RUN: perf record -e cycles -q -o %t.perf.data -- %t.exe 2> /dev/null
 
-CHECK-SPE-NO-LBR: PERF2BOLT: Starting data aggregation job
+RUN: (perf2bolt -p %t.perf.data -o %t.perf.boltdata --spe --nl %t.exe 2> /dev/null; exit 0) | FileCheck %s --check-prefix=CHECK-SPE-NO-LBR
 
-RUN: perf record -e cycles -q -o %t.perf.data -- %t.exe
-RUN: not perf2bolt -p %t.perf.data -o %t.perf.boltdata --spe %t.exe 2>&1 | FileCheck %s --check-prefix=CHECK-SPE-LBR
+RUN: (perf2bolt -p %t.perf.data -o %t.perf.boltdata --spe %t.exe 2> /dev/null; exit 0) | FileCheck %s --check-prefix=CHECK-SPE-LBR
 
-CHECK-SPE-LBR: PERF2BOLT: spawning perf job to read SPE branch events
+CHECK-SPE-NO-LBR: PERF2BOLT: spawning perf job to read SPE branch events (non-lbr)
+CHECK-SPE-LBR: PERF2BOLT: spawning perf job to read SPE brstack events
```

(the diff covers the entire test; lbr/non-lbr)

kaadam replied:
Thanks for clarifying this. Updated this test.

paschalis-mpeis commented on bolt/unittests/Profile/PerfSpeEvents.cpp:
```diff
@@ -88,6 +89,45 @@ struct PerfSpeEventsTestHelper : public testing::Test {
+    DA.BinaryMMapInfo.insert(std::make_pair(PID, MMap));
+
+    // Process buffer.
+    while (DA.hasData()) {
```
Would it be possible to call `parseBranchEvents` here instead of having a loop and replicating logic?

And then check any relevant buffers (ie `DA.BranchLBRs`  in a loop) ?
(probably no need to FallthroughLBRs at this point).

kaadam replied:
The main question is what we should test, 1) only the parser,  2) or more?

The `parseBranchEvents` function does a bit more thing than just parsing the output of the `perf script`. In `parseLBREntry` function, it is also recording/verifying the parsed traces, and determine the 'From/To' BinaryFunctions based on the sampled addresses. If we have a valid trace, the from/to pairs will be appended to `BranchLBRs` container. In our cases we use a dummy ELF file, so the BranchLBRs is empty.
1) If we would like to test only the parsing mechanism, we have to extract the proper logic for the testing.
2) If we would like to test the all complex logic, it needs to use a bit more complex input.

paschalis-mpeis replied:
Right. To do this we need to expand our context further, creating mock functions with valid addresses, allowing those branches to be parsed+stored in `BranchLBRs`.

That may be too much work; if that's the case we can skip.

paschalis-mpeis commented on bolt/unittests/Profile/PerfSpeEvents.cpp:
```diff
@@ -113,6 +153,37 @@ TEST_F(PerfSpeEventsTestHelper, SpeBranches) {
+                         "  1234  0xe001/0xe002/P/-/-/14/RET/-\n"
+                         "  1234  0xf001/0xf002/MN/-/-/8/COND/-\n";
+
+  LBREntry Entry1 = {0xa001, 0xa002, false};
+  LBREntry Entry2 = {0xb001, 0xb002, false};
+  LBREntry Entry3 = {0xc001, 0xc002, false};
+  LBREntry Entry4 = {0xd001, 0xd002, true};
+  LBREntry Entry5 = {0xe001, 0xe002, false};
+  LBREntry Entry6 = {0xf001, 0xf002, true};
+  std::vector<SmallVector<LBREntry, 2>> ExpectedSamples = {
+      {{Entry1}}, {{Entry2}}, {{Entry3}}, {{Entry4}}, {{Entry5}}, {{Entry6}},
+  };
```
nit: could expand directly the entries here without using the LBREntry variables

kaadam replied:
Yes, it could be. The reason is why I used the `LBREntry` type is to increasing the readability of the structure. Of course I agree to striving for simplicity.

kaadam replied:
Simplified, thanks for the hint.

paschalis-mpeis reviewed (commented):
Thanks a lot for your work Adam!
I commented on some changes and nits.

Also noting that for now this PR is stacked on top of #120741.

kaadam commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -180,13 +178,16 @@ void DataAggregator::start() {
-      errs() << "PERF2BOLT-ERROR: Arm SPE mode is combined only with "
-                "BasicAggregation.\n";
-      exit(1);
+      // pid    from_ip      to_ip        predicted?
+      // 12345  0x123/0x456/P/-/-/8/RET/-
```
Paschalis, thanks  for the review. yes, it's much better indeed.

kaadam committed:
Fix format issue

kaadam committed:
Fix typo

kaadam committed:
Removing dependency of the SPE BasicAggregation

This commit aim is to uncouple the SPE BRStack and SPE BasicAggregation approaches
based on the decision in issue #115333.

BRStack change relies on the unit test logic which was introduced by
Paschalis Mpeis (ARM) in #120741. Since it is a common part of the two aggregation
type technique, needs to retain an essential part of it.

All relevant tests to BasicAggregation is removed.

Co-Authored-By: Paschalis Mpeis <Paschalis.Mpeis@arm.com>

paschalis-mpeis commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -49,6 +49,11 @@ static cl::opt<bool>
                      cl::cat(AggregatorCategory));
 
+cl::opt<bool> ArmSPE("spe",
+                     cl::desc("Enable Arm SPE mode. Can combine with `--nl` "
+                              "to use in no-lbr mode"),
```
```suggestion
                     cl::desc("Enable Arm SPE mode."),
```

kaadam replied:
Done.

paschalis-mpeis commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -175,11 +180,16 @@ void DataAggregator::start() {
-    launchPerfProcess("events without LBR",
-                      MainEventsPPI,
+  if (opts::ArmSPE) {
+    // pid    from_ip      to_ip        predicted/missed not-taken?
```
Can simplify to:

```suggestion
    // pid    from_ip      to_ip      flags
```

and then expand 'flags' below:
- P/N: whether branch was Predicted or Mispredicted
- N: optionally appears when the branch was Not-Taken (ie fall-through)

kaadam replied:
Added the suggested details.

paschalis-mpeis commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -998,9 +1008,21 @@ ErrorOr<LBREntry> DataAggregator::parseLBREntry() {
-  if (MispredStr.size() != 1 ||
-      (MispredStr[0] != 'P' && MispredStr[0] != 'M' && MispredStr[0] != '-')) {
-    reportError("expected single char for mispred bit");
+  // SPE brstack mispredicted flags might be two characters long: 'PN' or 'MN'.
```
nit:
```suggestion
  // SPE brstack mispredicted flags might be up to two characters long: 'PN' or 'MN'.
```


Can add that 'N' is optional.

kaadam replied:
Done.

paschalis-mpeis commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -1513,9 +1535,11 @@ void DataAggregator::printBranchStacksDiagnostics(
-                     TimerGroupDesc, opts::TimeAggregator);
+  std::string BranchEventTypeStr =
+      !opts::ArmSPE ? "branch events" : "SPE branch events in LBR-format";
+  outs() << "PERF2BOLT: " << BranchEventTypeStr << "...\n";
```
nit: adding back the verb from the original code:

```suggestion
  outs() << "PERF2BOLT: parse " << BranchEventTypeStr << "...\n";
```

paschalis-mpeis commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -1560,10 +1585,17 @@ std::error_code DataAggregator::parseBranchEvents() {
+        errs()
+            << "PERF2BOLT-WARNING: all recorded samples for this binary lack "
+               "SPE brstack entries. Record profile with:"
+               "perf record arm_spe_0/branch_filter=1/";
```
```suggestion
               "perf record -e 'arm_spe_0/branch_filter=1/'";
```

kaadam replied:
Done.

paschalis-mpeis commented on bolt/test/perf2bolt/X86/perf2bolt-spe.test:
```diff
@@ -0,0 +1,9 @@
+## Check that Arm SPE mode is unavailable on X86.
+
+REQUIRES: system-linux,x86_64-linux
+
```
This test now is failing. Probably this line was accidentally removed?

https://github.com/llvm/llvm-project/blob/089cfd0cd427f10013feef835fe181aa0c0459f3/bolt/test/perf2bolt/AArch64/perf2bolt-spe.test#L5

paschalis-mpeis commented on bolt/unittests/Profile/PerfSpeEvents.cpp:
```diff
@@ -0,0 +1,144 @@
+  std::unique_ptr<ObjectFile> ObjFile;
+  std::unique_ptr<BinaryContext> BC;
+
+  /// Compare LBREntries
```
nit: Could add that: `@return true if LBREntries are equal.`

Also, could end comments with a dot.

paschalis-mpeis commented on bolt/unittests/Profile/PerfSpeEvents.cpp:
```diff
@@ -0,0 +1,144 @@
+      // Check the parsed LBREntries.
+      const auto *ActualIter = Sample.LBR.begin();
+      const auto *ExpectIter = ExpectedSamples[NumSamples].begin();
+      while (ActualIter != Sample.LBR.end() &&
+             ExpectIter != ExpectedSamples[NumSamples].end())
```
nit/tip: may be able to use [zip_equal](https://llvm.org/docs/ProgrammersManual.html#the-zip-functions) with structured binding here.

paschalis-mpeis reviewed (commented):
Hey Adam,

Thanks for addressing the comments!
`AArch64/perf2bolt-spe.test` seems to be failing. See comment below.
I've also added a few more nits and cleanups.

paschalis-mpeis commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -1560,10 +1585,17 @@ std::error_code DataAggregator::parseBranchEvents() {
-                "mode may be limited)\n";
+      if (!opts::ArmSPE)
+        errs()
+            << "PERF2BOLT-WARNING: all recorded samples for this binary lack "
```
When a user is on an older Linux/perf version, they may get zero samples.

Could you amend this message to note that too? Please also indicate which Linux version is required for support [1].

In those cases running `perf script -F pid,brstack --itrace=bl` only returns:
```
PID
1234
1234
1234
```

In other words, the brstack field is missing, in a silent failure.


---

[1] @Leo-Yan confirms that his brstack work landed on Linux Kernel v6.14. Thanks Leo!

kaadam replied:
I added a more explanation to error message based on your comment. I also mentioned the perf version in the comment.  Could you verify the brstack support will land in V6.14?

paschalis-mpeis replied:
Yes, that is correct.

kaadam committed:
Address reviewers 2

kaadam committed:
Simplifies SpeBranchesWithBrstack testcase

The test could be simplified after #143288 PR since
the validation phase is removed from parseLBRSample.
Now we can use branchLBRs container for the testing.
Formerly if Bolt was supplied with mock addresses, branchLBRs container
was empty due to validation phase.

paschalis-mpeis commented on bolt/unittests/Profile/PerfSpeEvents.cpp:
```diff
@@ -0,0 +1,119 @@
+
+    DA.parseBranchEvents();
+
+    EXPECT_EQ(DA.BranchLBRs.size(), SampleSize);
```
Thanks for simplifying the tests! 🙂 

Recent changes in bolt profile structures have now enabled this [previous comment](https://github.com/llvm/llvm-project/pull/129231#discussion_r2037463440).

I think we can re-introduce the loop, checking on each entry the mispreds and the taken counts. Just like what is done here for a particular entry (`1234, 0xd001, 0xd002`), but instead do it in a loop, for all `ExpectedSamples`.

paschalis-mpeis reviewed (commented):
Hey Adam,

Thanks for addressing the comments. A final thing would be to reintroduce the loop for checking the samples in the unit-test.

kaadam committed:
Added the suggested changes

kaadam committed:
Fixes typo

kaadam commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -1565,10 +1592,20 @@ std::error_code DataAggregator::parseBranchEvents() {
+               "Linux-perf is v6.14 or higher for brstack support. "
+               "With an older Linux-perf you may get zero samples. "
+               "Plese also make sure about you recorded profile with: "
+               "perf record -e 'arm_spe_0/branch_filter=1/'.";
```
Updated. Thanks.

kaadam committed:
clang format fix

kaadam committed:
Updated the test case.

paschalis-mpeis commented on bolt/unittests/Profile/PerfSpeEvents.cpp:
```diff
@@ -0,0 +1,167 @@
+namespace llvm {
+namespace bolt {
+
+/// Perform checks on perf SPE branch events combined with other SPE or perf
+/// events.
```
nit:
```suggestion
/// Perform checks on perf SPE branch events.
```

kaadam committed:
updated comments

paschalis-mpeis commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -1527,9 +1551,11 @@ void DataAggregator::printBranchStacksDiagnostics(
-  outs() << "PERF2BOLT: parse branch events...\n";
-  NamedRegionTimer T("parseBranch", "Parsing branch events", TimerGroupName,
-                     TimerGroupDesc, opts::TimeAggregator);
+  std::string BranchEventTypeStr =
+      !opts::ArmSPE ? "branch events" : "SPE branch events in LBR-format";
```
nit:

```suggestion
  std::string BranchEventTypeStr =
      opts::ArmSPE ? "SPE branch events" : "branch events";
```

paschalis-mpeis reviewed (approved):
Hey Adam, thank you for all this work, LGTM! 

Any comments from Meta? (cc: @maksfb, @aaupov)

kaadam committed:
Address reviewers 3

kaadam commented:
Paschalis, thanks for your review. Updated the all suggested change.

kaadam committed:
Adjusting SpeBranchesWithBrstack testcase to #143289

aaupov reviewed (approved):
LGTM, thank you for contributing this. 
Do you have plans to include an end-to-end (binary) test, with perf data with SPE samples?

kaadam committed:
Address reviewers 4

kaadam commented:
@aaupov @paschalis-mpeis Just one more thing. If you all find the change appropriate, may I ask one fo you to merge the PR at the final stage? I have no permission to do that. Thank you.
